hathor/util.py

```diff
@@ -503,3 +503,157 @@ def progress(iter_tx: Iterator['BaseTransaction'], *, log: Optional['structlog.s
     dt_total = LogDuration(t_final - t_start)
     tx_rate = '?' if dt_total == 0 else count / dt_total
     log.info('loaded', tx_count=count, tx_rate=tx_rate, total_dt=dt_total, height=h, blocks=block_count, txs=tx_count)
+
+
+class peekable(Iterator[T]):
+    """Adaptor class to peek what will be returned by next(iterator)
+
+
+    >>> it = peekable(range(10))
+    >>> iter(it) is it
+    True
+    >>> it.peek()
+    0
+    >>> next(it)
+    0
+    >>> next(it)
+    1
+    >>> it.peek()
+    2
+    >>> bool(it)
+    True
+    >>> next(it)
+    2
+    >>> list(it)
+    [3, 4, 5, 6, 7, 8, 9]
+    >>> bool(it)
+    False
+    >>> it.peek()
+    Traceback (most recent call last):
+    ...
+    ValueError: iterator was exhausted
+
+    """
+
+    def __init__(self, it: Iterable[T]) -> None:
+        self._it: Optional[Iterator[T]] = iter(it)
+        # XXX: using Optional[Tuple[T]] makes it so the iterator can yield None, and it would be correctly peekable,
+        #      which is different from not having a next element to peek into
+        self._head: Optional[Tuple[T]] = None
+
+    def _peek(self) -> Optional[Tuple[T]]:
+        if self._head is None and self._it is None:
+            return None
+        if self._head is None:
+            assert self._it is not None
+            try:
+                self._head = next(self._it),
+            except StopIteration:
+                self._it = None
+                return None
+        return self._head
+
+    def __iter__(self) -> Iterator[T]:
+        return self
+
+    def __next__(self) -> T:
+        if self._head is not None:
+            (x,), self._head = self._head, None
+            return x
+        elif self._it is not None:
+            try:
+                return next(self._it)
+            except StopIteration:
+                self.it = None
+                raise
+        else:
+            raise StopIteration()
+
+    def peek(self) -> T:
+        x = self._peek()
+        if x is None:
+            raise ValueError('iterator was exhausted')
+        y, = x
+        return y
+
+    def __bool__(self) -> bool:
+        return self._peek() is not None
+
+
+def _identity(x: T) -> T:
+    return x
+
+
+class sorted_merger(Iterator[T]):
+    """ Adaptor class to merge multiple sorted iterators into a single iterator that is also sorted.
+
+    Note: for this adaptor to work as expected the input iterators have to already be sorted, but if they aren't, the
```

Comment:
Yes, however not doing that was an explicit decision. Because of the limitation that UTXOs with either timelock or heightlock won't have the amount in sort order, I chose to make this not a HARD assumption, but a SOFT assumption. The end result won't be significantly affected, but we can no longer guarantee sorted order for the results when the results include timelock/heightlocks.

I think we should further decide what to do with the timelock/heightlock situation and revisit this when we have a decision.

Reply:
You're right. We shouldn't enforce it due to some anomalies caused by the way we are sorting UTXOs.